provider/adapters/src/fallback/adapter.rs

```diff
@@ -51,7 +51,8 @@ pub struct LocaleFallbackProvider<P> {
 impl<P> LocaleFallbackProvider<P>
 where
     P: DataProvider<LocaleFallbackLikelySubtagsV1Marker>
-        + DataProvider<LocaleFallbackParentsV1Marker>,
+        + DataProvider<LocaleFallbackParentsV1Marker>
+        + DynamicDataProvider<LocaleFallbackSupplementV1Marker>,
```

Comment:
Why is this dynamic?

Reply:
I arrived at this approach because:

1. The LocaleFallbackProvider wants to pre-load all fallback data
    - Why: Calling `for_config` is best as a cheap operation; it is invoked for every data key lookup, and we don't want to hit the data bundle and deserialize the fallback data each time.
2. The LocaleFallbackProvider should not eagerly link all fallback supplement keys
    - Why: There are going to be several components, starting with collation, that have fallback supplements. If an app is not linking `icu_collator`, they should not link the collation fallback supplement.
3. Therefore, we put a bound on a `DynamicDataProvider` and at runtime attempt to load the known fallback supplements during the LocaleFallbacker constructor (without adding the tagged path string to the binary).

Alternatives considered:

1. Add the locale fallback supplement key to KeyedDataMarker
    - Advantage: We can use a proper DataKey; metadata remains stack-only
    - Disadvantage: More tricky to bubble through the stack, and it doesn't solve the problem of leveraging a proper `DataProvider`
3. Add all the fallback data to the main fallback data key, with datagen generating different subsets of data in the main data key at datagen time
    - Advantage: Smaller impact on the runtime data provider model; don't need to rely on `DynamicDataProvider`
    - Disadvantage: Violates the property we currently maintain where a resource path uniquely identifies a resource (you may end up with multiple variants of the same resource with the same path); no automatic data slicing
3. Use a single key for supplemental data and leverage extension keywords to choose which variant
    - Advantage: Don't need to rely on `DynamicDataProvider`
    - Disadvantage: No automatic data slicing; unclear which extension keyword to use for this

The only known disadvantage of the proposed model is that if you're using a custom BakedDataProvider, you may need to add your own `impl_dynamic_data_provider` line for this to work, like I did in `icu_testdata` below. Sorry.

I'm willing to consider alternative 3 if you feel it's better.

Reply:
Hmm I don't think this is a great developer experience for anyone using databake. They'll have to bake and then know that if they're using collator fallbacking they need to use the macro to implement `DynamicDataProvider` for this one key. Actually even if they're *not* using collator, but are using fallback, they have to implement `DynamicDataProvider` for supplemental data, but with an error. I guess we can add this logic to databake, but alternative 3 sounds cleaner.

Reply:
Discussion: move forward with `fallback/supplement@1/und-x-collator.json`. Datagen should become smart enough to figure out which fallback supplements it needs to include.

Reply:
Updated discussion and conclusion:

- Start with option "A4" from the meeting notes: put the individual supplmenet DataProvider bounds on the fallback provider, and datagenshould generate empty databake DataProvider impls based on what is needed
- We could evolve into "A5" later: have code-size-optimized fallback provider constructors
- Try to re-use existing fallback supplement data structs when possible, but OK to add new ones, tailored for each component

Reply:
Options copied from [meeting notes](https://docs.google.com/document/d/1gKZCaiTG2eeyxrIKUF2tf6PIsaPCIBaUVRXD-JDh6dI/edit?pli=1#heading=h.782ztm2u1bwu)

 - A0 Shane: DynamicDataProvider (current impl)
 - A1 Shane: Add the locale fallback supplement key to KeyedDataMarker
 - A2 Shane: Add all the fallback data to the main fallback data key, with datagen generating different subsets of data in the main data key at datagen time
 - A3 Shane: Use a single key for supplemental data and leverage extension keywords to choose which variant
 - A4 Manish: Have all of the keys on the fallback provider constructor bound, but in datagen, make empty databake impls if they are not necessary
 - A5 Manish: Require the fallback provider constructor call site to specify which supplements to enable
```rust
fn try_new_generic1<P, M1>(p: &p) where
P: DataProvider<M>,
M: DataMarker<Yokeable = LocaleFallbackSupplementV1<'static>>
```
 

As mentioned by Shane above, the plan is to start with A4 and potentially add A5 later.